File: AndroidLab/ch6/src/main/java/com/example/ch6/section1/Test1_1Activity.java
package com.example.ch6.section1;

import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.SimpleAdapter;
import android.widget.Toast;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import com.example.ch6.R;
import com.example.ch6.databinding.ActivityTest11Binding;

import java.util.ArrayList;
import java.util.HashMap;

public class Test1_1Activity extends AppCompatActivity {

    ActivityTest11Binding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);

        binding = ActivityTest11Binding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });

        //데이터 획득..
        String[] arrayData = getResources().getStringArray(R.array.location);

        //항목 하나에 데이터 하나 출력...
        ArrayAdapter arrayAdapter
                = new ArrayAdapter(this, android.R.layout.simple_list_item_1, arrayData);
        binding.list1.setAdapter(arrayAdapter);
        //항목 선택 이벤트 추가..
        binding.list1.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            //position : 선택한 항목의 index...
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                Toast.makeText(Test1_1Activity.this, arrayData[position], Toast.LENGTH_SHORT)
                        .show();
            }
        });

        //SimpleAdapter..................... 한항목에 여러건의 데이터가 순차적으로..........
        //SimpleAdapter 에 데이터 구성이 ArrayList<Map<String, String>> 으로 준비..
        ArrayList<HashMap<String, String>> simpleData = new ArrayList<>();
        for(int i = 0; i < arrayData.length; i++){
            HashMap<String, String> map = new HashMap<>();
            map.put("name", arrayData[i]);
            map.put("content", String.valueOf(i * 10));
            simpleData.add(map);
        }

        SimpleAdapter simpleAdapter = new SimpleAdapter(
          this,
          simpleData,//항목을 구성하기 위한 데이터..
          android.R.layout.simple_list_item_2,//항목을 구성하기 위한 layout xml
          new String[]{"name", "content"},//항목의 데이터가 Map, Map 에서 데이터를 추출할 때 사용할 key
          new int[]{android.R.id.text1, android.R.id.text2}//layout xml 에 항목 데이터를 출력할 뷰의 id
        );
        binding.list2.setAdapter(simpleAdapter);
    }
}

File: Android3/ch3/src/main/java/com/example/ch3/section1/Test1_1Activity.kt
package com.example.ch3.section1

import android.os.Bundle
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.lifecycle.ViewModelProvider
import com.example.ch3.R
import com.example.ch3.databinding.ActivityTest11Binding

//ViewModel 이용 방법..................

class Test1_1Activity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()

        val binding = ActivityTest11Binding.inflate(layoutInflater)
        setContentView(binding.root)

        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        //case1 - ViewModel 객체 선언....
        //관심의 분리는 맞지만.. ViewModel 을 이용한 Activity 상태 유지는 불가능..
        //Activity 상태 유지가 되려면 ViewModel 의 lifecycle 이 Activity 와 다르게 되어야 하는데
        //코드에서 직접 생성함으로.. 매번 생성되어서..
//        val viewModel = MyViewModel1()

        //case2 - ViewModelProvider
        //직접생성하지 않았다..
        //관심 분리, 액티비티 상태 유지 가능하다..
        //아래의 코드에 의해 ViewModel 이 매번 생성되지는 않는다.. 싱글톤으로 운영한다..
//        val viewModel = ViewModelProvider(this).get(MyViewModel1::class.java)


        //case3 - delegate 이용....
        //ViewModelProvider 의 축약형이다.. 내부적으로.. ViewModelProvider 이용한다..
        val viewModel: MyViewModel1 by viewModels()

        binding.button.setOnClickListener {
            Toast.makeText(this, "${viewModel.data}, ${viewModel.someFun()}",
                Toast.LENGTH_SHORT).show()
        }
    }
}